x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutDatafeedAction.java

```diff
@@ -80,6 +84,7 @@ public TransportPutDatafeedAction(Settings settings, TransportService transportS
         this.datafeedConfigProvider = new DatafeedConfigProvider(client, xContentRegistry);
         this.jobConfigProvider = new JobConfigProvider(client, xContentRegistry);
         this.xContentRegistry = xContentRegistry;
+        clusterService.addListener(event -> clusterState = event.state());
```

Comment:
> I think we should set the initial state as well, so `clusterState = clusterService.state()` immediately above this line. I guess in reality it doesn't matter as it could be impossible not to get some other cluster state update before a new datafeed is put. But it seems best to declare the intent that `clusterState` should not be `null`.

I had to revert this change in all the three places as the constructor is apparently too early to fetch the cluster state and many tests are failing with the exception: "initial cluster state not set yet" thrown from ClusterApplierService::state.
Instead, I skip updating mappings if clusterState is null and log warning. Please LMK if you have other idea on how to solve this problem.

Reply:
Thanks.  This sounds like the best approach.

Reply:
Why not used the `clusterstate` passed as an argument to `masterOperation()` 

`clusterState` can change while this action is executing if `clusterService` calls the listener, so you could be in a situation where one code branch it taken based on current clusterstate then the state changes and that code branch is invalid.

Reply:
Thanks for this suggestion @davidkyle. It definitely makes sense for TransportPutJobAction and TransportPutDatafeedAction which are both subclasses of TransportMasterNodeAction.
However, I'm not sure what to do with TransportPutDataFrameAnalyticsAction which is HandledTransportAction and doesn't seem to have any handle on clusterState other than through ClusterService listener...

Reply:
You can grab the cluster state from the cluster service at any time. State is locally cached by `ClusterService` so it is a fast call. 

`ClusterState state = clusterService.state();` 

Can be called in `doExecute` or where ever needed

Reply:
Don't take the state in the constructor though as the object may be constructed long before `doExecute` is called, which I guess is why you got the _initial cluster state not set yet_ error.

Reply:
Done.